File: src/data/courses/dsa/arrays-and-hashing/advanced-hashing/first-missing-positive/code/cpp/first_missing_positive_sol2.cpp
class Solution {
public:
    // Optimal: O(n) time, O(1) space — Cyclic sort: place each num at index num-1
    int firstMissingPositive(vector<int>& nums) {
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            // Keep swapping until nums[i] is in correct position or out of range
            while (nums[i] > 0 && nums[i] <= n && nums[nums[i]-1] != nums[i]) {
                swap(nums[i], nums[nums[i]-1]);
            }
        }
        // First index where nums[i] != i+1 is the answer
        for (int i = 0; i < n; i++) {
            if (nums[i] != i + 1) return i + 1;
        }
        return n + 1;
    }
};

File: src/data/courses/dsa/arrays-and-hashing/advanced-hashing/first-missing-positive/code/python/first_missing_positive_sol2.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        # Optimal: O(n) time, O(1) space — Cyclic sort
        n = len(nums)
        for i in range(n):
            while 1 <= nums[i] <= n and nums[nums[i]-1] != nums[i]:
                nums[nums[i]-1], nums[i] = nums[i], nums[nums[i]-1]
        for i in range(n):
            if nums[i] != i + 1: return i + 1
        return n + 1